### ISSUE TITLE
StopAsync in worker service's Worker class not invoked when running as Windows service at system shutdown/restart

### ISSUE THREAD

**Issue by lonewolfnsp** [aspnet/AspNetCore#17506]
wonder if you guys have encountered this before. I wanted to have grpc server+web api running as windows service as it's meant for a long runningprocess. The web API is basically just for runtime monitoring of internal states.. So I created a new grpc service (.net core 3), and add in the worker class, and add the controller etc.. and used the Microsoft.Extensions.Hosting.WindowsServices nuget package to enable Windows service. I used "sc create binpath= start= auto" to install it, then used "sc start " to start it.. observing the log file, I can see it's started, accessing the swagger page, I can see it's running fine.

I'm logging something to the log file when the StartAsync and StopAsync are called in the worker class and I can see that when the service starts, StartAsync is called, as it should be.

Now the funny thing is this, when the system is restarted, I don't see StopAsync being called..

in DOS, if I used dotnet .exe to run it, StartAsync is called, pressing CTRL+C to stop it, StopAsync is called.. but why is StopAsync not called when it's running as windows service and system is restarted? is this a bug in the windwos service nuget package itself? or am I missing something here?

Funny thing is, in Startup class, I even used IHostApplicationLifetime to intercept the ApplicationStarting, ApplicationStopping, but again, starting is called, but stopping is not called.. why?

can someone please enlighten me?

**Comment by Tratcher** [comment]
Duplicate of https://github.com/aspnet/Extensions/pull/2705, already addressed in 5.0.

**Comment by lonewolfnsp** [comment]
Sorry for asking but is there some kind of workaround I can use for now? coz 5.0 is going to take a year to show up. 

Is it an issue with the windows service nuget package or the .net core framework itself? I know in 2.x, an asp.net core web app hosted in IIS can get notified of system shutdown and application startup upon first incoming request via the IApplicationLifetime interface.. 

Just an unrelated issue, not sure if it's an issue or something I did wrongly. In asp.net core 2.x hosted in IIS, I found that it does not auto start with system startup.. I've to trigger the first incoming request to get it to start, it does not auto start byself.. is it still the case for 3.0? or am I missing something here?
